ecl/hql/hqlopt.cpp

```diff
@@ -3988,6 +3984,36 @@ IHqlExpression * CTreeOptimizer::doCreateTransformed(IHqlExpression * transforme
                     }
                     break;
                 }
+            case no_if:
+                {
+                    if (!isShared(child) && transformed->queryChild(1)->isConstant())
+                        return swapIntoIf(transformed);
+                    break;
+                }
+             }
+             break;
+         }
+    case no_select:
+        {
+            node_operator childOp = child->getOperator();
+            switch(childOp)
+            {
+                case no_if:
+                {
+                    if (!isShared(child) && child->isDatarow())
+                    {
+                        IHqlExpression * cond = child->queryChild(0);
+                        IHqlExpression * field = transformed->queryChild(1);
+                        OwnedHqlExpr newLeft = createNewSelectExpr(LINK(child->queryChild(1)), LINK(field));
+                        OwnedHqlExpr newRight = createNewSelectExpr(LINK(child->queryChild(2)), LINK(field));
+                        incUsage(newLeft);
+                        incUsage(newRight);
+                        OwnedHqlExpr newIf = createIf(LINK(cond), LINK(newLeft), LINK(newRight));
+                        newIf.setown(child->cloneAllAnnotations(transformed));
```

Comment:
I think this line is likely to be wrong, and killed my machine when running the regression tests.